pkg/worker/worker.go

```diff
@@ -0,0 +1,177 @@
+	// exited or was terminated by a signal
+	ExitCode int
+	// Exited reports whether the program has exited
+	Exited bool
+}
+
+// Worker defines the basic operations to manage Jobs.
+type Worker interface {
+	// Start creates a Linux process.
+	//    - command: command to be executed
+	// It returns the job ID and the execution error encountered.
+	Start(command Command) (jobID string, err error)
+	// Stop a running Job which kills a running process.
+	//    - ID: Job identifier
+	// It returns the execution error encountered.
+	Stop(jobID string) (err error)
+	// Query a Job to check the current status.
+	//    - ID: Job identifier
+	// It returns process status and the execution error
+	// encountered.
+	Query(jobID string) (status Status, err error)
+	// Streams the process output.
+	//    - ctx: context to cancel the log stream
+	//    - ID: Job identifier
+	// It returns read chan to stream process stdout/stderr and the
+	// execution error encountered.
+	Stream(ctx context.Context, jobID string) (logchan chan string, err error)
+}
+
+// NewWorker creates a new Worker instance.
+func NewWorker(config conf.Config) Worker {
+	return &worker{
+		logger: log.NewLogger(config),
+		jobs:   make(map[string]*Job),
+	}
+}
+
+// worker implementation.
+type worker struct {
+	// logger is responsible to handle the
+	// stdout and stderr of a running process
+	logger log.Logger
+	// jobs is concurrency safe map to store
+	// the requested jobs
+	jobs map[string]*Job
+	// mtx to control jobs concurret access
+	mtx sync.RWMutex
+}
+
+// Start runs a Linux single command with arguments.
+// If the command runs with success, a Job identifier will be returned.
+// A log file will be created with the Job id name to capture the stdout and stderr
+// of a running process.
+// To get the process status, the Job request will be stored in memory,
+// and a goroutine will be launched to update the job status when the process is finished.
+func (w *worker) Start(command Command) (jobID string, err error) {
+	cmd := exec.Command(command.Name, command.Args...)
+	jobID = uuid.NewString()
+	logfile, err := w.logger.Create(jobID)
+	if err != nil {
+		return
+	}
+	// redirect the stdout and stderr to the log file
+	cmd.Stdout = logfile
+	cmd.Stderr = logfile
+	if err = cmd.Start(); err != nil {
+		w.logger.Remove(jobID)
+		return
+	}
+	// create and store the job
+	w.mtx.Lock()
+	w.jobs[jobID] = &Job{ID: jobID, Cmd: cmd, Status: &Status{Pid: cmd.Process.Pid}}
+	w.mtx.Unlock()
+	// update the job status in background after
+	// the process exit
+	go w.updateJob(jobID)
+	return
+}
+
+// Stop terminates a running Job gracefully sending a SIGTERM to the process.
+// If the job doesn't exitis an error will be returned.
+func (w *worker) Stop(jobID string) (err error) {
+	w.mtx.RLock()
+	defer w.mtx.RUnlock()
+	job, ok := w.jobs[jobID]
+	if !ok {
+		return errors.New("Job not found")
+	}
+	return job.Cmd.Process.Signal(syscall.SIGTERM)
```

Comment:
If the job stops on its own at the same time, this will attempt to send a signal to a terminated process.

Reply:
To address it I can send the signal 0 to check if the process still running.